src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp

```diff
@@ -1478,3 +1482,61 @@ void ShenandoahFullGC::phase4_compact_objects(ShenandoahHeapRegionSet** worker_s
 
   heap->clear_cancelled_gc(true /* clear oom handler */);
 }
+
+void ShenandoahFullGC::phase5_restore_generation_usage() {
+  GCTraceTime(Info, gc, phases) time("Phase 5: Restore generation accounts", _gc_timer);
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  size_t num_regions = heap->num_regions();
+  size_t young_usage = 0;
+  size_t young_regions = 0;
+  size_t young_humongous_waste = 0;
+  size_t old_usage = 0;
+  size_t old_regions = 0;
+  size_t old_humongous_waste = 0;
+  ShenandoahHeapRegion* r;
+
+  for (size_t i = 0; i < num_regions; i++) {
+    switch (heap->region_affiliation(i)) {
+      case ShenandoahRegionAffiliation::FREE:
+        break;
+      case ShenandoahRegionAffiliation::YOUNG_GENERATION:
+        r = heap->get_region(i);
+        young_regions++;
+        young_usage += r->used();
+        if (r->is_humongous()) {
```

Comment:
Two nits:

1. There is code in the two arms here that can be consolidated into a work method that can then be used to update the appropriate young or old metric.
2. It's conceptually simpler to check if is_humongous() and then add size of object if is_humongous_start. You can there also do arithmetic for the waste in the last humongous region for that object. Then you can just skip all the is_humongous_cont() regions, and save some time.

Other than that, this looks good to me.